test/functional/lua/with_spec.lua

```diff
@@ -0,0 +1,292 @@
+    eq(buf2, val)
+  end)
+
+  it('does not cause ml_get errors with invalid visual selection', function()
+    exec_lua [[
+      local api = vim.api
+      local t = function(s) return api.nvim_replace_termcodes(s, true, true, true) end
+      api.nvim_buf_set_lines(0, 0, -1, true, {"a", "b", "c"})
+      api.nvim_feedkeys(t "G<C-V>", "txn", false)
+      vim._with({buf = api.nvim_create_buf(false, true)}, function() vim.cmd "redraw" end)
+    ]]
+  end)
+
+  it('can be nested crazily with hidden buffers', function()
+    eq(
+      true,
+      exec_lua([[
+      local function scratch_buf_call(fn)
+        local buf = vim.api.nvim_create_buf(false, true)
+        vim.api.nvim_set_option_value('cindent', true, {buf = buf})
+        return vim._with({buf = buf}, function()
+          return vim.api.nvim_get_current_buf() == buf
+            and vim.api.nvim_get_option_value('cindent', {buf = buf})
+            and fn()
+      end) and vim.api.nvim_buf_delete(buf, {}) == nil
+    end
+
+    return scratch_buf_call(function()
+      return scratch_buf_call(function()
+        return scratch_buf_call(function()
+          return scratch_buf_call(function()
+            return scratch_buf_call(function()
+              return scratch_buf_call(function()
+                return scratch_buf_call(function()
+                  return scratch_buf_call(function()
+                    return scratch_buf_call(function()
+                      return scratch_buf_call(function()
+                        return scratch_buf_call(function()
+                          return scratch_buf_call(function()
+                            return true
+                          end)
+                        end)
+                      end)
+                    end)
+                  end)
+                end)
+              end)
+            end)
+          end)
+        end)
+      end)
+    end)
+    ]])
+    )
+  end)
+
+  it('can return values by reference', function()
+    eq(
+      { 4, 7 },
+      exec_lua [[
+      local val = {4, 10}
+      local ref = vim._with({ buf = 0}, function() return val end)
+      ref[2] = 7
+      return val
+    ]]
+    )
+  end)
+end)
+
+describe('vim._with {win = }', function()
+  it('does not trigger autocmd', function()
+    exec_lua [[
+      local old = vim.api.nvim_get_current_win()
+      vim.cmd("new")
+      local new = vim.api.nvim_get_current_win()
+      vim.api.nvim_create_autocmd( { 'BufEnter', 'BufLeave', 'BufWinEnter', 'BufWinLeave' }, {
+        callback = function() _G.n = (_G.n or 0) + 1 end
+      })
+      vim._with({win = old}, function()
+      end)
+      assert(_G.n == nil)
+    ]]
+  end)
+
+  it('trigger autocmd if changed within context', function()
+    exec_lua [[
+      local old = vim.api.nvim_get_current_win()
+      vim.cmd("new")
+      local new = vim.api.nvim_get_current_win()
+      vim.api.nvim_create_autocmd( { 'BufEnter', 'BufLeave', 'BufWinEnter', 'BufWinLeave' }, {
```

Comment:
Here testing for `WinEnter` and `WinLeave` is more important. I'd either add those or replace `BufEnter` and `BufLeave`.